test/test_nn.py

```diff
@@ -8572,5 +8607,109 @@ def _buildEquivalentAffineTransforms3d(device, input_size, output_size, angle_ra
 # end TestNN.test_affine_* helpers
 
 
+# The following are some helpers for TestNN.test_transformer_number_match
+class DataBatch:
+    "Object for holding a batch of data with mask during training."
+    def __init__(self, src, trg=None, pad=0):
+        self.src = src.transpose(0, 1).contiguous()
+        self.src_mask = (src != pad).unsqueeze(-2)
+        if trg is not None:
+            self.trg = trg[:, :-1].transpose(0, 1).contiguous()
+            self.trg_y = trg[:, 1:].transpose(0, 1).contiguous()
+            self.ntokens = (self.trg_y != pad).data.sum().item()
+
+class NoamOpt:
+    "Optim wrapper that implements rate."
+    def __init__(self, model_size, factor, warmup, optimizer):
+        self.optimizer = optimizer
+        self._step = 0
+        self.warmup = warmup
+        self.factor = factor
+        self.model_size = model_size
+        self._rate = 0
+
+    def step(self):
+        self._step += 1
+        rate = self.rate()
+        for p in self.optimizer.param_groups:
+            p['lr'] = rate
+        self._rate = rate
+        self.optimizer.step()
+
+    def rate(self, step=None):
+        if step is None:
+            step = self._step
+        return self.factor * (self.model_size ** (-0.5) * 
+                              min(step ** (-0.5), step * self.warmup ** (-1.5)))
+
+class LabelSmoothingLoss(nn.Module):
+    "Implement label smoothing."
+    def __init__(self, size, padding_idx, smoothing=0.0):
+        super(LabelSmoothingLoss, self).__init__()
+        self.criterion = nn.KLDivLoss(size_average=False)
+        self.padding_idx = padding_idx
+        self.confidence = 1.0 - smoothing
+        self.smoothing = smoothing
+        self.size = size
+        self.true_dist = None
+
+    def forward(self, x, target):
+        assert x.size(1) == self.size
+        true_dist = x.data.clone()
+        true_dist.fill_(self.smoothing / (self.size - 2))
+        true_dist.scatter_(1, target.data.unsqueeze(1), self.confidence)
+        true_dist[:, self.padding_idx] = 0
+        mask = torch.nonzero(target.data == self.padding_idx)
+        if mask.dim() > 0:
+            true_dist.index_fill_(0, mask.squeeze(), 0.0)
+        self.true_dist = true_dist
+        return self.criterion(x, Variable(true_dist, requires_grad=False))
+
+def train_epoch(data_iter, model, criterion, opt):
```

Comment:
Since this is a common name, it's important to be local in scope to the test.

Reply:
please remove before finalizing the PR (see my comments above)